packages/dd-trace/test/scope/async_local_storage.spec.js

```diff
@@ -0,0 +1,46 @@
+'use strict'
+
+const {
+  AsyncLocalStorage,
+  AsyncResource,
+  executionAsyncId
+} = require('async_hooks')
+const Scope = require('../../src/scope/async_local_storage')
+const Span = require('opentracing').Span
+const testScope = require('./test')
+
+wrapIt()
+
+if (AsyncLocalStorage) {
+  describe('Scope (AsyncLocalStorage)', test)
+} else {
+  describe.skip('Scope (AsyncLocalStorage)', test)
+}
+
+function test () {
+  let scope
+  let span
+
+  beforeEach(() => {
+    scope = new Scope()
+    span = new Span()
+  })
+
+  it('should not break propagation for nested resources', done => {
+    scope.activate(span, () => {
+      const asyncResource = new AsyncResource(
+        'TEST', { triggerAsyncId: executionAsyncId(), requireManualDestroy: false }
+      )
+
+      asyncResource.runInAsyncScope(() => {})
+
+      expect(scope.active()).to.equal(span)
+
+      // AsyncLocalStorage context persists through `done()` unless we tell it
+      // not to. Without this, the following tests will run inside this scope.
+      scope._storage.exit(done)
```

Comment:
What will happen if the test fails before this line runs?

Reply:
It just wouldn't reach the `done` call. 🤷 

I assume the test suite has a timeout system built-in, which would have been initialized outside the context and therefore not be impacted.

Reply:
Unfortunately this is not the case, or at least I don't think that's the case from the behavior I've seen in the past. There were many issues with this in the past, which is why `wrapIt` was created. However, I'm not sure how difficult supporting multiple scope implementations would be with `wrapIt`. Since multiple implementations should never be used at the same time, maybe a parameter could be added to pick a specific one?

Reply:
Just tested a simulated fail there with a throw and no `done` call. Seems to work fine.